src/dmd/identifier.d

```diff
@@ -166,75 +154,53 @@ nothrow:
      */
     extern (D) static Identifier generateIdWithLoc(string prefix, const ref Loc loc)
     {
-        import dmd.root.filename: absPathThen;
-
-        // see below for why we use absPathThen
-        return loc.filename.absPathThen!((absPath)
+        // generate `<prefix>_L<line>_C<col>`
+        OutBuffer idBuf;
+        idBuf.writestring(prefix);
+        idBuf.writestring("_L");
+        idBuf.print(loc.linnum);
+        idBuf.writestring("_C");
+        idBuf.print(loc.charnum);
+
+        /**
+         * Make sure the identifiers are unique per filename, i.e., per module/mixin
+         * (`path/to/foo.d` and `path/to/foo.d-mixin-<line>`). See issues
+         * https://issues.dlang.org/show_bug.cgi?id=16995
+         * https://issues.dlang.org/show_bug.cgi?id=18097
+         * https://issues.dlang.org/show_bug.cgi?id=18111
+         * https://issues.dlang.org/show_bug.cgi?id=18880
+         * https://issues.dlang.org/show_bug.cgi?id=18868
+         * https://issues.dlang.org/show_bug.cgi?id=19058
+         */
+        static struct Key { Loc loc; string prefix; }
+        __gshared uint[Key] counters;
+
+        static if (__traits(compiles, counters.update(Key.init, () => 0u, (ref uint a) => 0u)))
         {
-
-            // this block generates the "regular" identifier, i.e. if there are no collisions
-            OutBuffer idBuf;
-            idBuf.writestring(prefix);
-            idBuf.writestring("_L");
-            idBuf.print(loc.linnum);
-            idBuf.writestring("_C");
-            idBuf.print(loc.charnum);
-
-            // This block generates an identifier that is prefixed by the absolute path of the file
-            // being compiled. The reason this is necessary is that we want unique identifiers per
-            // module, but the identifiers are generated before the module information is available.
-            // To guarantee that each generated identifier is unique without modules, we make them
-            // unique to each absolute file path. This also makes it consistent even if the files
-            // are compiled separately. See issues:
-            // https://issues.dlang.org/show_bug.cgi?id=16995
-            // https://issues.dlang.org/show_bug.cgi?id=18097
-            // https://issues.dlang.org/show_bug.cgi?id=18111
-            // https://issues.dlang.org/show_bug.cgi?id=18880
-            // https://issues.dlang.org/show_bug.cgi?id=18868
-            // https://issues.dlang.org/show_bug.cgi?id=19058.
-            OutBuffer fullPathIdBuf;
-
-            if (absPath)
-            {
-                // replace characters that demangle can't handle
-                for (auto ptr = absPath; *ptr != '\0'; ++ptr)
+            // 2.082+
+            counters.update(Key(loc, prefix),
+                () => 1u,          // insertion
+                (ref uint counter) // update
```

Comment:
I had to use `ref` here, although it makes no sense in this case. To be fair, I haven't tried returning void and incrementing the ref directly.